text/0517-io-os-reform.md

```diff
@@ -447,7 +457,182 @@ counts, arguments to `main`, and so on).
+  already-read data together with an error. For writers, the main
+  change is to make `write` only perform a single underlying write
+  (returning the number of bytes written on success), and provide a
+  separate `write_all` method.
+
+* **Parsing/serialization**. The `Reader` and `Writer` traits
+  currently provide a large number of default methods for
+  (de)serialization of various integer types to bytes with a given
+  endianness. Unfortunately, these operations pose atomicity problems
+  as well (e.g., a read could fail after reading two of the bytes
+  needed for a `u32` value).
+
+  Rather than complicate the signatures of these methods, the
+  (de)serialization infrastructure is removed entirely -- in favor of
+  instead eventually introducing a much richer
+  parsing/formatting/(de)serialization framework that works seamlessly
+  with `Reader` and `Writer`.
+
+  Such a framework is out of scope for this RFC, but the
+  endian-sensitive functionality will be provided elsewhere
+  (likely out of tree).
+
+* **The error type**. The traits currently use `IoResult` in their
+  return types, which ties them to `IoError` in particular. Besides
+  being an unnecessary restriction, this type prevents `Reader` and
+  `Writer` (and various adapters built on top of them) from moving to
+  `libcore` -- `IoError` currently requires the `String` type.
+
+  With associated types, there is essentially no downside in making
+  the error type generic.
+
+With those general points out of the way, let's look at the details.
+
+### Nonatomic results
+[Nonatomic results]: #nonatomic-results
+
+To clarity dealing with nonatomic operations and improve their
+ergonomics, we introduce some new types into `std::error`:
+
+```rust
+// The progress so far (T) paired with an err (Err)
+struct PartialResult<T, Err>(T, Err);
+
+// An operation that may fail after having made some progress:
+// - S is what's produced on complete success,
+// - T is what's produced if an operation fails part of the way through
+type NonatomicResult<S, T, Err> = Result<S, PartialResult<T, Err>>;
+
+// Ergonomically throw out the partial result
+impl<T, Err> FromError<PartialResult<T, Err>> for Err { ... }
+```
+
+The `NonatomicResult` type (which could use a shorter name)
+encapsulates the common pattern of operations that may fail after
+having made some progress. The `PartialResult` type then returns the
+progress that was made along with the error, but with a `FromError`
+implementation that makes it trivial to throw out the partial result
+if desired. For example, the following would be expected to compile:
+
+```rust
+fn write(buf: &[u8]) -> NonatomicResult<(), uint, Error> { /* ... */ }
+
+fn write_bytes() -> Result<(), Error> {
+    try!(write!(&[1, 2, 3, 4]));
+    Ok(())
+}
+```
+
+### `Reader`
+[Reader]: #reader
+
+The updated `Reader` trait (and its extension) is as follows:
+
+```rust
+trait Reader {
+    type Err; // new associated error type
```

Comment:
This would shadow anything named `Error` in the surrounding scope for impls, which could make it a little awkward when there are types named `Error` due to [RFC 356](https://github.com/rust-lang/rfcs/blob/master/text/0356-no-module-prefixes.md). That said, we used `Result` for [RFC 439](https://github.com/rust-lang/rfcs/blob/master/text/0439-cmp-ops-reform.md), so we probably should spell out `Error` and just live with the need for type aliases in some cases.